### PR TITLE
Fix Bugzilla 24582 - Detect unsafe `cast(bool[])`

### PR DESCRIPTION
**Pull request by ntrel**
Deprecate in safe code.

### CODE REVIEW

**Comment by thewilsonator** [comment]
Dependant PR merged

**Comment by dkorpel** [comment]
Like I mentioned on bugzilla, I don't see how these casts are unsafe.

**Comment by ntrel** [comment]
@dkorpel I added a comment there with an example:
https://issues.dlang.org/show_bug.cgi?id=24582#c3

Only the runtime array cast is unsafe, the other 2 are OK - thanks. I need to update this pull to allow the literal cast, as well as change the error to a deprecation.

**Comment by RazvanN7** [comment]
Why is the cast being deprecated? Wouldn't it make more sense to simply fix the literal cast to create an array of 0 and 1? For the sake of consistency, casting a literal or a variable shouldn't yields different results.

**Comment by dkorpel** [comment]
> Why is the cast being deprecated?

It's not, it's disallowed in `@safe` code.

> Wouldn't it make more sense to simply fix the literal cast to create an array of 0 and 1s?

That's what happens.

> For the sake of consistency, casting a literal or a variable shouldn't yield different results.

Currently it does, but changing that behavior is a breaking change and out of scope for this bug report.

**Comment by ntrel** [comment]
> Wouldn't it make more sense to simply fix the literal cast to create an array of 0 and 1s

That would go against the spec:

https://dlang.org/spec/expression.html#cast_array
>  The cast is done as a type paint

So the raw data is not modified.

**Comment by dkorpel** [comment]
> No, as I posted on bugzilla, the byte contents is unchanged:

For the runtime cast yes, but the literal cast is different:
```D
immutable c = cast(bool[]) [2, 4]; // literal cast applies to each element
pragma(msg, c); // [true, true]
```

**Comment by ntrel** [comment]
@dkorpel OK, sorry - so we are in agreement!

**Review by dkorpel — APPROVED** [review]

File: compiler/test/fail_compilation/bool_cast.d [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/*
+REQUIRED_ARGS: -de
+TEST_OUTPUT:
+---
+fail_compilation/bool_cast.d(12): Deprecation: cast from `ubyte[]` to `bool[]` not allowed in safe code
```
Comment (thewilsonator) [inline compiler/test/fail_compilation/bool_cast.d:5]: (Not that you need to do this right away, but) can we link to/refer users to the spec section/other official documentation, as to _why_ this is an unsafe operation? It was certainly not immediately obvious to me as to why this is the case.

File: compiler/src/dmd/expressionsem.d [hunk 2]
```diff
@@ -15286,6 +15296,9 @@ Expression resolveLoc(Expression exp, const ref Loc loc, Scope* sc)
 
     Expression visitStructLiteral(StructLiteralExp exp)
     {
+        if (!exp.elements)
+            return exp;
```
Comment (ntrel) [inline compiler/src/dmd/expressionsem.d:15300]: These snuck in accidentally; will fix bit later.

File: compiler/test/fail_compilation/bool_cast.d (hunk 1 above)
Comment (ntrel) [reply compiler/test/fail_compilation/bool_cast.d]: I've added a supplemental message for now.

**Comment by ntrel** [comment]
Tests now passing (other than 1 interop network request failure)

**Pull request merged**